### ISSUE TITLE
Add an option to display only attribute combinations that exist and have stock greater than zero

### ISSUE THREAD

**Issue by mariannk** [nopSolutions/nopCommerce#16]
We already have a similar product option implemented ("Allow only existing attribute combinations"). It does not allow to add non-existing combinations to the cart. And this work item is just to display/hide such combination in the public store BEFORE a customer click "add to cart" (client-side validation). 

Also note that it can cause some issues when you have several attributes. Once choosing one of attributes, it can hide other attributes which actually have some stock with some other value of the first attribute 

please note that we already support display of stock levels for attribute combinations (dynamic) 

Also see  https://github.com/nopSolutions/nopCommerce/issues/2909 (example of UI implementation)

**Comment by Opsis** [comment]
This one is very important for one of my leads to migrate away from magento to nopcommerce.
As posted on the forum: http://www.nopcommerce.com/boards/t/44919/track-inventory-by-product-attributes.aspx

**Comment by cklenk** [comment]
Is there a good reason why this is marked as won't fix? Is there a plugin that does this?

**Comment by colinwphillips** [comment]
I had already created a plugin that does this, please let me know if I should share code.
I added a line to PrepareProductAttributeModels, under:
`if (attribute.ShouldHaveValues())
{`
`var attributeValues = _productAttributeService.GetProductAttributeValues(attribute.Id);`
I added:
`FilterAttributes(product, ref attributeValues);`

Implementation:
`private static void FilterAttributes(Product product, ref IList<ProductAttributeValue> attributeValues)
        {
            if (product == null || attributeValues == null) return;
            attributeValues = (from attr in attributeValues let id = attr.Id let match = product.ProductAttributeCombinations.FirstOrDefault(a => a.AttributesXml.Contains($"<Value>{id}</Value>")) where match != null select attr).ToList();
        }`

**Comment by miguelgimenez** [comment]
Hi Collin,

I would appreciate it if you could share your code for this.

Regards,

Miguel

**Comment by zamliyconsulting** [comment]
> I had already created a plugin that does this, please let me know if I should share code.
> I added a line to PrepareProductAttributeModels, under:
> `if (attribute.ShouldHaveValues()) {`
> `var attributeValues = _productAttributeService.GetProductAttributeValues(attribute.Id);`
> I added:
> `FilterAttributes(product, ref attributeValues);`
> 
> Implementation:
> `private static void FilterAttributes(Product product, ref IList<ProductAttributeValue> attributeValues) { if (product == null || attributeValues == null) return; attributeValues = (from attr in attributeValues let id = attr.Id let match = product.ProductAttributeCombinations.FirstOrDefault(a => a.AttributesXml.Contains($"<Value>{id}</Value>")) where match != null select attr).ToList(); }`

Tried your code, it does nothing, unfortunately.

**Comment by cklenk** [comment]
> Nothing indeed. We moved away from nopcommerce since 3.7

Hi @stephdep I'm curious what solution are you using now?

**Comment by AR1ES** [comment]
As a solution, you can change the code or create a new plugin and override the ProductModelFactory.

```
public class ProductModelFactory : Nop.Web.Factories.ProductModelFactory
{
	protected override IList<ProductDetailsModel.ProductAttributeModel> PrepareProductAttributeModels(Product product, ShoppingCartItem updatecartitem) {
		var model = base.PrepareProductAttributeModels(product, updatecartitem);
		foreach(var productAttributeModel in model) {
			productAttributeModel.Values = productAttributeModel.Values
				.Where(m => product.ProductAttributeCombinations.Any(c => c.StockQuantity > 0 && 
					c.AttributesXml.Contains($"ID=\"{productAttributeModel.Id}\"") &&
					c.AttributesXml.Contains($"<Value>{m.Id}</Value>")
					)).ToList();
		}
		return model;
	}
}
```

**Comment by arieldeil** [comment]
Many thanks!

It should also effect product attributes display in filters. 
We can hide products from being displayed when out of stock but when using combinations (which are basically many different products combined) we do not have the same feature. This makes users frustrating when they are filtering for a specific color or size but get "out of stock" messages, or worse, the size doesn't exists in product page due to current feature.

**Comment by holydk** [comment]
Closed #16

**Comment by saeidmh83** [comment]
> As a solution, you can change the ProductModelFactory class code or create a new plugin and override the ProductModelFactory:
> 
> ```cs
> public class ProductModelFactory : Nop.Web.Factories.ProductModelFactory
> {
> 	protected override IList<ProductDetailsModel.ProductAttributeModel> PrepareProductAttributeModels(Product product, ShoppingCartItem updatecartitem) {
> 		var model = base.PrepareProductAttributeModels(product, updatecartitem);
> 		foreach(var productAttributeModel in model) {
> 			productAttributeModel.Values = productAttributeModel.Values
> 				.Where(m => product.ProductAttributeCombinations.Any(c => c.StockQuantity > 0 && 
> 					c.AttributesXml.Contains($"ID=\"{productAttributeModel.Id}\"") &&
> 					c.AttributesXml.Contains($"<Value>{m.Id}</Value>")
> 					)).ToList();
> 		}
> 		return model;
> 	}
> }
> ```
This solution worked for me. Thank you
